### ISSUE TITLE
Type Error in Combination with Discriminating Union

### ISSUE THREAD

**Issue by fabb** [microsoft/TypeScript#35700]
**TypeScript Version:**  3.7.2

**Search Terms:** discriminating union, types incompatible, not assignable

**Expected behavior:**

The code should compile.

**Actual behavior:**

**Related Issues:** None found yet.

**Code**

Note that this is a reduced example. My real-world use case uses the `schema-dts` json-ld library which provides the `Vehicle`/`Car`/etc. types.

```ts
type Vehicle =
    | ({
          '@type': 'Vehicle'
      } & {
          speed?: number
      })
    | ({
          '@type': 'Car'
      } & {
          speed?: number
      } & {
          roofLoad?: number
      })

let type = 'Car' as 'Vehicle' | 'Car'

// TypeScript error: Type '{ '@type': "Vehicle" | "Car"; }' is not assignable to type 'Vehicle'. Types of property ''@type'' are incompatible. Type '"Vehicle"' is not assignable to type '"Car"'.
const v: Vehicle = {
    '@type': type,
}

```

Interestingly, modifying the `Vehicle` type would make it compile (but I cannot do that in my real-world project because the types are provided by the `schema-dts` json-ld library):

```ts
type Vehicle =
    | ({
          '@type': 'Vehicle'
          speed?: number
      })
    | ({
          '@type': 'Car'
          speed?: number
          roofLoad?: number
      })

let type = 'Car' as 'Vehicle' | 'Car'

// TypeScript error: Type '{ '@type': "Vehicle" | "Car"; }' is not assignable to type 'Vehicle'. Types of property ''@type'' are incompatible. Type '"Vehicle"' is not assignable to type '"Car"'.
const v: Vehicle = {
    '@type': type,
}

```

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/#code/C4TwDgpgBAahAWBLAxgG2gXgFBV1APlABQDeOeFuA5AAKiRUBcUVcSaEV5FAvlAGRQylSgGdIEACYB+ZgDsArgFsARhABO3PDwCUWwqS0ja9TsyoBhAIbquIvoOEi84iFNlRFqjUagOhvhTqAPbBAGYAMsFWMvLKapr2eljowFCmUBgs1rZQVqIsbCjoVATZNlxYAPRVUAAq4BAAysjqiGBpGiHqzA2QLCQsdI1MUABERRxjZWM5YwDcfqWIBXLBafmiiADmclYq6OnB6Y2FCMWcAHT1jQXhUGAhkOqgLCYjpTbQiHLIwUpgKzARAHCDXPrQKgTc5TZardZ5URbXb7Q7AY4ZKFzKiXLB-OSiNIAN2Yk0OWSc1GGDGYpgANFgeFggA)

**Comment by fabb** [comment]
Same error on 3.8.0-dev.20191216

**Comment by rbuckton** [comment]
I believe this has been addressed in the time between when this issue was filed and now, as I can no longer repro it as of 3.8.3.

**Comment by rbuckton** [comment]
I believe this case was addressed by #36663
